Multi Pressure Advance (Nozzle relation)

ianalexis opened the pull request:
# Multi PA

Enable the possibility of loading more than one PA according to the nozzle, as it has a direct relationship.

<img width="707" height="241" alt="imagen" src="https://github.com/user-attachments/assets/20597fd8-11ed-4756-a2a0-7dc67c469284" />

Useful for avoiding multiple profiles just for nozzle changes, but also for new machines with tool changes that allow for nozzle changes while maintaining the material.

- "Legacy” value with a single value and no nozzle, take that value as before.
- Single value (e.g., 0.4,0.2), it is taken as legacy.
- If there is a value without a incomplete value ("0.5," || ",0.2") is ignored.
- If nozzle does not match the listed values:
  - 1 PA value = using that value
  - Multi PA values = ~~linear~~ PChip interpolation like Adaptive Pressure Advance (IF WE WANT TO GO BACK TO LINEAR WE MUST REVERT THE LAST COMMIT).
- "Smart" Order, if the user set pa/nozzle the algorithm will swap the values
  - If 1 column only contain multiples of 0.05 and:
    - 2 column only contain multiples of 0.05 = No Swap
    - 2 column contain any non-multiple of 0.05 = No Swap
      - Nozzle is the one with more 0.05 multiple values.
  - If 1 column contain any non-multiple of 0.05 and:
    - 2 column only contain multiples of 0.05 = Swap
    - 2 column contain non-multiple of 0.05:
      - Nozzle is the one with more 0.05 multiple values.

Nowdays Prusa do this by:
```json
    "filament_start_gcode": [
        "M572 S{if nozzle_diameter[filament_extruder_id]==0.4}0.02
			{elsif nozzle_diameter[filament_extruder_id]==0.5}0.018
			{elsif nozzle_diameter[filament_extruder_id]==0.6}0.015
			{elsif nozzle_diameter[filament_extruder_id]==0.8}0.01
			{elsif nozzle_diameter[filament_extruder_id]==0.25}0.09
			{elsif nozzle_diameter[filament_extruder_id]==0.3}0.065
			{else}0{endif} ; Filament gcode\nM142 S40 ; set heatbreak target temp"
    ],
```

+240 Prusa profiles that could be converted into 41 profiles analyzed in a logarithmic table.
<img width="1284" height="1006" alt="imagen" src="https://github.com/user-attachments/assets/f755fffc-4c1f-4411-bfbc-1263cf680d26" />

## Adaptive Pressure Advance:
I just adapt when it takes the PA value but maybe we have to:

- add the nozzle size to it as a value = Nozzle, PA, flow, acceleration 
  - Using multiple APA models.
  - interpolate it's resulting values with a potential curve based in the PA.
- use it as it is and do not adapt it

> [!NOTE]
> Added PA and APA to filament_options_with_variant so you can set it by extrusor.

## TODO after Merge

- Prusa profiles could be readapted to use this structure and no longer have the values in filament_start_gcode.

ianalexis committed:
Nozzle - Pa

SoftFever commented:
I like this idea. Would you like to take on the challenge of improving the UI so we can dynamically add new rows to specify PA for different Nozzle Diameter values? Something like this:

<img width="355" height="121" alt="Screenshot 2025-10-19 at 10 15 03 PM" src="https://github.com/user-attachments/assets/07df4aa9-958a-44b0-b54f-2e4025f5eb6f" />

ianalexis committed:
Fix PA with: Set PA at start

Co-Authored-By: Rodrigo Faselli <162915171+RF47@users.noreply.github.com>

ianalexis committed:
Fix

ianalexis committed:
Smart Order

ianalexis committed:
Compact

ianalexis committed:
Optimization

coryrc commented on src/libslic3r/GCode.cpp:
```diff
@@ -6705,6 +6710,96 @@ float GCode::interpolate_value_across_layers(float start_value, float end_value)
+    int first_col_multiples_count = 0;
+    int second_col_multiples_count = 0;
+
+    for (const auto& [first, second] : pa_values) {
```
If you're keeping this, I think the following would be safer:

```
   for (const auto& [first, second] : pa_values) {
        auto left = is_multiple_of_005(first),
          right = is_multiple_of_005(second);
        if (left && !right)  first_col_multiples_count++;
        if (!left && right) second_col_multiples_count++;
    }

    // Swap columns if needed (second column has more nozzle-like values)
    if (second_col_multiples_count > first_col_multiples_count && first_col_multiples_count == 0) {
        for (auto& [first, second] : pa_values) {
            std::swap(first, second);
        }
    }
```

That way we only swap if the second column contains *only* nozzle-like dimensions.

ianalexis replied:
Did it, take a look if you like it

coryrc commented on src/libslic3r/PrintConfig.cpp:
```diff
@@ -2136,13 +2136,19 @@ void PrintConfigDef::init_fff_params()
+                     "Set as: Nozzle Size, Pressure Advance Value.\n"
+                     "Multiple values can be set. For example\n"
+                     "0.25,0.09 \n0.3,0.065 \n0.35,0.05 \n0.4,0.02 \n0.5,0.018 \n0.6,0.012 \n0.8,0.01 \n"
+                     "Missing values will be calculated with linear interpolation.");
```
, or the value for the smallest/biggest nozzle if smaller/bigger than all the listed nozzles.

coryrc commented on src/slic3r/GUI/Field.cpp:
```diff
@@ -757,7 +764,7 @@ void TextCtrl::BUILD() {
     temp->SetLabel(_L(m_opt.sidetext));
 	auto text_ctrl = m_opt.multiline ? (wxTextCtrl *)temp : ((TextInput *) temp)->GetTextCtrl();
-    text_ctrl->SetLabel(text_value);
+    text_ctrl->SetValue(text_value);
```
why change?

ianalexis commented on src/slic3r/GUI/Field.cpp:
```diff
@@ -736,6 +736,13 @@ void TextCtrl::BUILD() {
 		const ConfigOptionStrings *vec = m_opt.get_default_value<ConfigOptionStrings>();
 		if (vec == nullptr || vec->empty()) break; //for the case of empty default value
 		text_value = vec->get_at(m_opt_idx);
+		// For multiline fields, unescape newlines and other escape sequences
```
In coStrings, the default value may contain escape sequences  like \n. That block calls unescape_string_cstyle to convert those sequences into actual charcters when the field is multiline

ianalexis commented on src/libslic3r/GCode.cpp:
```diff
@@ -6705,6 +6710,96 @@ float GCode::interpolate_value_across_layers(float start_value, float end_value)
+    // Early return for single value - no interpolation needed
+    if (pa_values.size() == 1) return pa_values[0].second;
+
+    // Auto-detect if columns are swapped by checking if values are multiples of 0.05 (nozzle characteristic)
```
It was because I myself encountered the error of writing the values backwards. For now, I prefer this approach.

ianalexis committed:
Coryfication

Co-Authored-By: coryrc <github@corycross.org>

ianalexis commented on src/libslic3r/GCode.cpp:
```diff
@@ -6705,6 +6710,96 @@ float GCode::interpolate_value_across_layers(float start_value, float end_value)
 }
 
+// Calculate the correct pressure advance value based on nozzle diameter
+double GCode::get_pressure_advance_for_extruder(unsigned int filament_id) const
```
Hmm, I don't know if I'm going to do it in this PR.

ianalexis committed:
Fix PA reset to use correct extruder value

ianalexis committed:
Wiki fix

ianalexis committed:
Use PCHIP interpolation for pressure advance values

Replaces the previous binary search and linear interpolation logic in get_pressure_advance_for_extruder with a monotonic PCHIP interpolator for smoother and more accurate results. Deduplicates nozzle diameter entries to avoid zero-length segments and falls back to linear interpolation if PCHIP construction fails.

ianalexis committed:
Orca tag

coryrc commented:
As an aside, the rebase pushes make it hard for me to keep track of what changes are being made.

ianalexis committed:
Cache pressure_advance value before set/reset

Replace duplicate calls to get_pressure_advance_for_extruder with a local pressure_advance variable in multiple places in GCode.cpp (filament change, initial PA write, and set_extruder paths). This ensures the same PA value is used for both writer.set_pressure_advance and m_pa_processor->resetPreviousPA, reduces redundant calls, and improves readability and consistency.

ianalexis committed:
Update Plater.cpp

ianalexis committed:
Preserve legacy pressure_advance in placeholders

Create an effective DynamicConfig (copying config_override if present) and inject a scalar "pressure_advance" ConfigOptionFloats using get_pressure_advance_for_extruder(current_filament_id) before calling the placeholder parser. This preserves legacy behavior where [pressure_advance] resolves to a scalar value while still passing other overrides into the parser.

CooperGerman commented:
Hi, just came across this PR as I was searching a way to have a single filament profile for all my printers but have different PA values according to which one is used and which nozzle size.

My two first thoughts were :
- Update the PA field in Orca from a single value to a nozzle (or in my case machine) dependent value as you are implementing.
- Store material dependant configs such as pressure advance (maybe retractions ?) as part of the Spoolman profile that could later be loaded into Orca through [this PR](https://github.com/OrcaSlicer/OrcaSlicer/pull/4771).

The major point in favor of this to me is to keep a small amount of filament profiles as I would have to create a parent profile and then specific machine related profiles. In my case on filament would have :
`1 parent + 4 (different toolheads) * 5 (nozzle sizes) = 21` children profiles hence that many json files to keep track of.

I understand this PR aims at fixing part of my concern right ?

ianalexis commented:
> I understand this PR aims at fixing part of my concern right ?

TL;DR:
Something like that.
1 profile with all nozzle/pa relations under each extrusor tab (in cases with different extrusors like the new BBL printer).

With this pr you should only need 1 profile per filament.
If 2 extrusor types (like that new BBL with DD and Bowden) appears you should only create 2 tabs in the same filament.

1. Tab to DD and all nozzle/pa relations for DD.
2. Tab for Bowden wit all it's nozzle/pa relations.

When i started this PR the main goal was to avoid creating 3 profiles for the same printer just for when i switched nozzles.
Then i found Prusa's logic and i hated it hahahah.
Now with multi tool printers like the U1 and orca's ability to mix nozzle sizes in the same project.
Clone the profile just for PA change? NO!
Add a table with each relation and then use the corresponding to each.

PS: Maybe i should add something like that for retraction override. Have to test it with a real printer to check if that's something needed.

ianalexis committed:
PA values multi extruder compatible and specific

SoftFever committed:
Merge branch 'main' into multiple-pa

Fuzzylog2x commented:
Where are we at with this?  Presently (v2.5.0 nightly), the PA values entered for a standard flow nozzle affect the high flow nozzle, and vice versa.  In my testing, the PA values for these nozzle types are different.

<img width="2050" height="506" alt="image" src="https://github.com/user-attachments/assets/1fb9cb14-df82-4dd3-8be1-2dfb10f12f94" />

ianalexis commented:
> Where are we at with this? Presently (v2.5.0 nightly), the PA values entered for a standard flow nozzle affect the high flow nozzle, and vice versa. In my testing, the PA values for these nozzle types are different.

we are thinking in another implementation for this due of how multitool printers profiles evolved.

Fuzzylog2x commented:
> > Where are we at with this? Presently (v2.5.0 nightly), the PA values entered for a standard flow nozzle affect the high flow nozzle, and vice versa. In my testing, the PA values for these nozzle types are different.
> 
> we are thinking in another implementation for this due of how multitool printers profiles evolved.

Why not simply add some additional fields to the comma delimited table?  For instance:

PA, Flow, Accel, Type, Diameter, Drive Type (optional)

0.020, 25,00, 2000, Standard, 0.4
0.025, 15.00, 2000, Standard, 0.4
0.030, 5.000, 2000, Standard, 0.4
0.022, 25,00, 2000, High Flow, 0.4
0.027, 15.00, 2000, High Flow, 0.4
0.032, 5.000, 2000, High Flow, 0.4
0.105, 25,00, 2000, High Flow, 0.4, Bowden
0.125, 15.00, 2000, High Flow, 0.4, Bowden
0.150, 5.000, 2000, High Flow, 0.4, Bowden
